From 3d285ec9b1798749fcf08b2ffd1332887f6e64d3 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 17 Mar 2025 14:03:12 +0800
Subject: [PATCH 209/236] mtk: hostapd: include vendor element in MBSSID
 nontransmitted BSSID profile

Since the RSNO setting might be different between TX BSS and non-TX
BSS(es), RSNO should be included in nontransmitted BSSID profile.
Meanwhile, since RSNO is Vendor Element, other Vendor Elements should
also be include in nontransmitted BSSID profile.

Here we only add common Vendor element:
1. WMM
2. MBO
3. RSNO

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/beacon.c     | 15 ++++++---------
 src/ap/beacon.h     |  6 ++++++
 src/ap/ieee802_11.c | 20 ++++++++++++++++++++
 3 files changed, 32 insertions(+), 9 deletions(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index f2567c32a..54e7d8511 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -403,8 +403,7 @@ static u8 * hostapd_get_wpa_ie(struct hostapd_data *hapd, u8 *pos, size_t len)
 }
 
 
-static u8 * hostapd_get_rsne_override(struct hostapd_data *hapd, u8 *pos,
-				      size_t len)
+u8 * hostapd_get_rsne_override(struct hostapd_data *hapd, u8 *pos, size_t len)
 {
 	const u8 *ie;
 
@@ -417,8 +416,7 @@ static u8 * hostapd_get_rsne_override(struct hostapd_data *hapd, u8 *pos,
 }
 
 
-static u8 * hostapd_get_rsne_override_2(struct hostapd_data *hapd, u8 *pos,
-					size_t len)
+u8 * hostapd_get_rsne_override_2(struct hostapd_data *hapd, u8 *pos, size_t len)
 {
 	const u8 *ie;
 
@@ -431,8 +429,7 @@ static u8 * hostapd_get_rsne_override_2(struct hostapd_data *hapd, u8 *pos,
 }
 
 
-static u8 * hostapd_get_rsnxe_override(struct hostapd_data *hapd, u8 *pos,
-				       size_t len)
+u8 * hostapd_get_rsnxe_override(struct hostapd_data *hapd, u8 *pos, size_t len)
 {
 	const u8 *ie;
 
@@ -445,7 +442,7 @@ static u8 * hostapd_get_rsnxe_override(struct hostapd_data *hapd, u8 *pos,
 }
 
 
-static size_t hostapd_get_rsne_override_len(struct hostapd_data *hapd)
+size_t hostapd_get_rsne_override_len(struct hostapd_data *hapd)
 {
 	const u8 *ie;
 
@@ -456,7 +453,7 @@ static size_t hostapd_get_rsne_override_len(struct hostapd_data *hapd)
 }
 
 
-static size_t hostapd_get_rsne_override_2_len(struct hostapd_data *hapd)
+size_t hostapd_get_rsne_override_2_len(struct hostapd_data *hapd)
 {
 	const u8 *ie;
 
@@ -467,7 +464,7 @@ static size_t hostapd_get_rsne_override_2_len(struct hostapd_data *hapd)
 }
 
 
-static size_t hostapd_get_rsnxe_override_len(struct hostapd_data *hapd)
+size_t hostapd_get_rsnxe_override_len(struct hostapd_data *hapd)
 {
 	const u8 *ie;
 
diff --git a/src/ap/beacon.h b/src/ap/beacon.h
index e390929df..53bd3c6f4 100644
--- a/src/ap/beacon.h
+++ b/src/ap/beacon.h
@@ -76,4 +76,10 @@ u8 * hostapd_unsol_bcast_probe_resp(struct hostapd_data *hapd,
 				    struct unsol_bcast_probe_resp *ubpr);
 void hostapd_gen_per_sta_profiles(struct hostapd_data *hapd);
 
+u8 * hostapd_get_rsne_override(struct hostapd_data *hapd, u8 *pos, size_t len);
+u8 * hostapd_get_rsne_override_2(struct hostapd_data *hapd, u8 *pos, size_t len);
+u8 * hostapd_get_rsnxe_override(struct hostapd_data *hapd, u8 *pos, size_t len);
+size_t hostapd_get_rsne_override_len(struct hostapd_data *hapd);
+size_t hostapd_get_rsne_override_2_len(struct hostapd_data *hapd);
+size_t hostapd_get_rsnxe_override_len(struct hostapd_data *hapd);
 #endif /* BEACON_H */
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index fc7a68a21..9307dccbc 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -9008,6 +9008,16 @@ static size_t hostapd_eid_mbssid_elem_len(struct hostapd_data *hapd,
 				bss, NULL, true, false);
 #endif /* CONFIG_IEEE80211BE */
 
+		/* WFA vendor elements which cannot be inherited if one of them
+		 * is different from that of TxBSS.
+		 * TODO: Check if any vendor elem is different or should not be inherited at all
+		 */
+		nontx_profile_len += hostapd_eid_wmm_len(bss);
+		nontx_profile_len += hostapd_mbo_ie_len(bss);
+		nontx_profile_len += hostapd_get_rsne_override_len(bss);
+		nontx_profile_len += hostapd_get_rsne_override_2_len(bss);
+		nontx_profile_len += hostapd_get_rsnxe_override_len(bss);
+
 		if (ie_count)
 			nontx_profile_len += 4 + ie_count + 1;
 
@@ -9182,6 +9192,16 @@ static u8 * hostapd_eid_mbssid_elem(struct hostapd_data *hapd, u8 *eid, u8 *end,
 			eid = hostapd_eid_eht_basic_ml_common(bss, eid, NULL,
 							      true, false);
 #endif /* CONFIG_IEEE80211BE */
+		/* WFA vendor elements which cannot be inherited if one of them
+		 * is different from that of TxBSS.
+		 * TODO: Check if any vendor elem is different or should not be inherited at all
+		 */
+		eid = hostapd_eid_wmm(bss, eid);
+		eid = hostapd_eid_mbo(bss, eid, end - eid);
+		eid = hostapd_get_rsne_override(bss, eid, end - eid);
+		eid = hostapd_get_rsne_override_2(bss, eid, end - eid);
+		eid = hostapd_get_rsnxe_override(bss, eid, end - eid);
+
 		if (ie_count) {
 			*eid++ = WLAN_EID_EXTENSION;
 			*eid++ = 2 + ie_count + 1;
-- 
2.45.2

